core/camel-base/src/main/java/org/apache/camel/impl/converter/TypeResolverHelper.java

```diff
@@ -0,0 +1,192 @@
+     * For instance: From String.class, To: int.class (would match a method such as myConverter(String, Integer) or
+     * myConverter(String, int).
+     *
+     * @param  typeConvertible the type converter pair
+     * @param  converters      the map of all known converters
+     * @return                 the type converter or null if unknown
+     */
+    static TypeConverter tryAssignableFrom(
+            TypeConvertible<?, ?> typeConvertible, Map<TypeConvertible<?, ?>, TypeConverter> converters) {
+
+        /*
+         Let's try classes derived from this toType: basically it traverses the entries looking for assignable types
+         matching both the "from type" and the "to type" which are NOT Object (we usually try this later).
+         */
+        final Optional<Map.Entry<TypeConvertible<?, ?>, TypeConverter>> first = converters.entrySet().stream()
+                .filter(v -> v.getKey().isAssignableMatch(typeConvertible))
+                .findFirst();
+
+        return first.map(Map.Entry::getValue).orElse(null);
+    }
+
+    /**
+     * Try to resolve a TypeConverter by looking at Array matches for a given "from" type. This also includes evaluating
+     * candidates matching a receiver of Object[] type.
+     *
+     * @param  typeConvertible the type converter pair
+     * @param  converters      the map of all known converters
+     * @return                 the type converter or null if unknown
+     */
+    static TypeConverter tryObjectArrayConverters(
+            TypeConvertible<?, ?> typeConvertible, Map<TypeConvertible<?, ?>, TypeConverter> converters) {
+
+        return tryObjectArrayConverters(typeConvertible.getFrom(), typeConvertible.getTo(), converters);
+    }
+
+    private static TypeConverter tryObjectArrayConverters(
+            Class<?> fromType, Class<?> toType, Map<TypeConvertible<?, ?>, TypeConverter> converters) {
+        // Let the fallback converters handle the primitive arrays
+        if (fromType.isArray() && !fromType.getComponentType().isPrimitive()) {
+            /* We usually define our converters are receiving an object array (Object[]), however, this won't easily match:
+             * because an object array is not an interface or a super class of other array types (i.e.: not a super class
+             * of String[]). So, we take the direct road here and try check for an object array converter right away.
+             */
+            return converters.get(new TypeConvertible<>(Object[].class, toType));
+        }
+
+        return null;
+    }
+
+    /**
+     * Try to resolve the TypeConverter by forcing a costly and slow recursive check.
+     *
+     * @param  typeConvertible the type converter pair
+     * @param  converters      the map of all known converters
+     * @return                 the type converter or null if unknown
+     */
+    static TypeConverter tryMatch(
+            TypeConvertible<?, ?> typeConvertible, Map<TypeConvertible<?, ?>, TypeConverter> converters) {
```

Comment:
All this stream filter stuff does have a bit of overhead in the JDK - sometimes code with just a for loop is faster.

Reply:
I'm going to give it a try with a plain for loop and see how it goes.

Reply:
There's indications that the plain loop is better as it creates less objects and all that. So, I changed to a loop as you suggested.